### DISCUSSION TITLE
scope["scheme"] should be passable for proxy support alongside with scope["root_path"] for redirect_slashes to work property

### DISCUSSION THREAD

**Discussion by alexey-pelykh** [fastapi/fastapi#12484] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
N/A
```

### Description

Exposing a FastAPI via a HTTPS-terminating load balancer causes `redirect_slashes` to emit a redirect to HTTP location. It would be great to be able to pass scope["scheme"] so that this code in Starlette will kick in:

https://github.com/encode/starlette/blob/46131a1af875a5be2190d91713a43ee80c8311c6/starlette/routing.py#L760

https://github.com/encode/starlette/blob/46131a1af875a5be2190d91713a43ee80c8311c6/starlette/datastructures.py#L33

### Operating System

Linux

### FastAPI Version

0.115.2

### Pydantic Version

N/A

### Python Version

N/A

**Comment by alexey-pelykh** [comment]
Plus it is a security concern. Consider the following case (assuming proxy/load-balancer do the 307 HTTP->HTTPS redirects and clients follows redirects and allows both HTTP and HTTPS outgoing requests):

```
> POST https://example.com/myendpoint + sensitive data
< 307 redirect to http://example.com/myendpoint/ [redirect_slashes kick in]
> POST http://example.com/myendpoint/ + sensitive data [here's were security risk is]
< 307 redirect to https://example.com/myendpoint/
> POST https://example.com/myendpoint/ + sensitive data
```

**Comment by yakubka** [comment]
The `scope["scheme"]` value is set by the ASGI server based on the actual connection protocol (http or https). When your application runs behind a TLS-terminating reverse proxy, the proxy connects to the backend over plain HTTP, so `scope["scheme"]` is `"http"` even though the original client used HTTPS.

The standard way to fix this is to use Starlette`s `ProxyHeadersMiddleware`, which reads the `X-Forwarded-Proto` header set by the proxy and updates `scope["scheme"]` accordingly.

```python
from starlette.middleware.trustedhost import TrustedHostMiddleware
from uvicorn.middleware.proxy_headers import ProxyHeadersMiddleware

app.add_middleware(ProxyHeadersMiddleware, trusted_hosts="*")
```

After adding this middleware, `request.url.scheme` and any code that reads `scope["scheme"]` will return `"https"` when the proxy header is present.

Be careful with `trusted_hosts="*"` in production. If untrusted clients can send `X-Forwarded-Proto: https`, they can spoof the scheme. Set `trusted_hosts` to the IP address or hostname of your proxy to prevent spoofing.
